semantic-convention-utils/src/main/java/org/hypertrace/semantic/convention/utils/http/HttpSemanticConventionUtils.java

```diff
@@ -196,4 +257,257 @@ public static boolean isValidUrl(String url) {
+    for (String path : URL_PATH_ATTRIBUTES) {
+      if (attributeValueMap.get(path) != null) {
+        String s = attributeValueMap.get(path).getValue();
+        if (StringUtils.isNotBlank(s) && s.startsWith(SLASH)) {
+          return getPathFromUrlObject(s).map(str -> removeTrailingSlash(str));
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpMethod(Event event) {
+
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String method : METHOD_ATTRIBUTES) {
+      if (attributeValueMap.get(method) != null
+          && !StringUtils.isBlank(attributeValueMap.get(method).getValue())) {
+        return Optional.of(attributeValueMap.get(method).getValue());
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpScheme(Event event) {
+
+    Optional<String> url = getHttpUrlUtil(event);
+    if (url.isPresent() && isAbsoluteUrl(url.get())) {
+      try {
+        return Optional.of(getNormalizedUrl(url.get()).getProtocol());
+      } catch (MalformedURLException e) {
+        e.printStackTrace();
+      }
+    }
+    return getHttpSchemeUtil(event);
+  }
+
+  private static Optional<String> getHttpSchemeUtil(Event event) {
+
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String scheme : SCHEME_ATTRIBUTES) {
+      if (attributeValueMap.get(scheme) != null
+          && !StringUtils.isEmpty(attributeValueMap.get(scheme).getValue())) {
+        return Optional.of(attributeValueMap.get(scheme).getValue());
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpUrl(Event event) {
+    Optional<String> url = getHttpUrlUtil(event);
+    if (url.isPresent() && !isAbsoluteUrl(url.get())) {
+      return Optional.empty();
+    }
+    return getHttpUrlUtil(event);
+  }
+
+  //  input url to populateurlparts
+  private static Optional<String> getHttpUrlUtil(Event event) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String url : FULL_URL_ATTRIBUTES) {
+      if (attributeValueMap.get(url) != null) {
+        String urlVal = attributeValueMap.get(url).getValue();
+        if (!StringUtils.isBlank(urlVal) && isValidUrl(urlVal)) {
+          if (isAbsoluteUrl(urlVal)) {
+            return Optional.of(urlVal);
+          } else {
+            Optional<String> url2 = getHttpUrlForOTelFormat(attributeValueMap);
```

Comment:
why is this return under top level check fo urlVal?

Reply:
This is in reference to this function:
https://github.com/hypertrace/hypertrace-ingester/blob/0261e7eef2641b0576c6cba05a215e976b6676f5/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L567

It will enter this if else block only when a matching key is present in the attributemap. This means that the url was set earlier using addvaluetobuilder method. Now if it is absolute url, we directly return without overwriting the url

Reply:
But `maybeSetHttpUrlForOtelFormat` is called after processing all the attributes